### ISSUE TITLE
Support for webpack?

### ISSUE THREAD

**Issue by hanfeisun** [videojs/videojs-contrib-hls#600]
In the version of 1.x videojs-contrib-hls, I can shim the package like this:

```
import vjs  from 'video.js';
window.videojs = vjs;

require('imports?this=>window!videojs-contrib-hls');

```

However, in 2.0.1 videojs-contrib-hls, this method doesn't work. 

A dependency called `webworkify` will throw an error

```
Uncaught TypeError: Cannot convert undefined or null to object
module.exports  @   index.js:10
VirtualSourceBuffer @   virtual-source-buffer.js:66
addSourceBuffer @   html-media-source.js:168
setupSourceBuffer_  @   videojs-contrib-hls.js?./~/imports-loader?this=>window:621
handleSourceOpen    @   videojs-contrib-hls.js?./~/imports-loader?this=>window:532
data.dispatcher @   video.js:18626
trigger @   video.js:18730
EventTarget.trigger @   video.js:7950

```

Will videojs-contrib-hls plan to support webpack in the future?

**Comment by dmlap** [comment]
I think this was the issue behind https://github.com/videojs/mux.js/pull/53. That was fixed in mux.js >= 2.0.1, which you should use if you build this project yourself (and in our next release when that happens). Do you see the same error if you build contrib-hls yourself?

**Comment by samward1985** [comment]
+1

**Comment by samward1985** [comment]
Eh - no luck... @hanfeisun did you find a solution?

**Comment by hanfeisun** [comment]
@samward1985  No, it doesn't work under Webpack now.. So I am still using `videojs-contrib-hls 1.3.11`..

**Comment by andrewmnlv** [comment]
Hello all!)
It works for me:
   resolve: {
      alias: {
         'videojs-contrib-hls': __dirname + '/node_modules/videojs-contrib-hls/dist/videojs-contrib-hls
      }
   }

videojs-contrib-hls v2.2.0

**Comment by gkatsev** [comment]
I think contrib-hls 2.2.0 won't require any aliases anymore. Though, unfortunately, that version also requires videojs 5.10.1 and up.

**Comment by andrewmnlv** [comment]
@gkatsev i have videojs v5.10.2, but webworkify breaks webpack build because of
[no fallbacks for arguments](https://github.com/substack/webworkify/blob/master/index.js#L3)

**Comment by gkatsev** [comment]
If webworkify is the only problem, we should consider swapping it out for another module with similar functionality that works in both webpack and browserify.

**Comment by gkatsev** [comment]
There seems to be a webworkify-webpack module which may work for webpack but it doesnt work for browserify. Someone should write a module that works for both.

**Comment by gkatsev** [comment]
After some thought and some discussion in babel's chatroom, it seems like the correct solution is to not rely on a packer transform or loader to the web worker inline but rather do it as a separate source transform. I saw https://github.com/mohayonao/inline-worker which may help but also maybe there's some babel plugin that allows us to do this.
Unfortunately, this is a low priority for us because there's more pressing specific issues and it can be worked around by using the dist file or maybe running webworkify-webpack over the code. If anyone is willing to submit a PR to fix this issue we'll definitely take a look and work with you to land it.
Thanks!

**Comment by ScottLNorvell** [comment]
I was having the same issue with webpack and contrib-hls. I was able to get it working by aliasing webworkify to webworkify-webpack in my webpack config like so:

```javascript
// ...
resolve: {
  alias: {
    webworkify$: 'webworkify-webpack'
  }
},
// ...
``` 
The strange caveat is that I need point to version 1.0.6 in my package.json instead of the current version (which is 1.1ish). Got that little tidbit from a separate issue [here](https://github.com/mapbox/mapbox-gl-js/issues/1649#issuecomment-222819121)

Hope this helps anyone!

**Comment by nehakhandelwal** [comment]
I am facing the same and added below code too.

resolve: {
  alias: 
{ 'videojs-contrib-hls': __dirname + '/node_modules/videojs-contrib-hls/dist/videojs-contrib-hls', 
    webworkify$: 'webworkify-webpack'
  }
},
 and getting this error
19:31:03.939 TypeError: _videoJs2.default is undefined
[72]</<() bundle.js%20line%205005%20%3E%20eval:8312
[72]<() bundle.js%20line%205005%20%3E%20eval:7848
s() bundle.js%20line%205005%20%3E%20eval:7
s/<() bundle.js%20line%205005%20%3E%20eval:7
[71]</<() bundle.js%20line%205005%20%3E%20eval:7668
[71]<() bundle.js%20line%205005%20%3E%20eval:7640
s() bundle.js%20line%205005%20%3E%20eval:7
s/<() bundle.js%20line%205005%20%3E%20eval:7
[76]</<() bundle.js%20line%205005%20%3E%20eval:8886
[76]<() bundle.js%20line%205005%20%3E%20eval:8870
s() bundle.js%20line%205005%20%3E%20eval:7
s/<() bundle.js%20line%205005%20%3E%20eval:7
[94]</<() bundle.js%20line%205005%20%3E%20eval:14668
[94]<() bundle.js%20line%205005%20%3E%20eval:14627
s() bundle.js%20line%205005%20%3E%20eval:7
e() bundle.js%20line%205005%20%3E%20eval:7
<anonymous> bundle.js%20line%205005%20%3E%20eval:7
<anonymous> bundle.js%20line%205005%20%3E%20eval:7
<anonymous> bundle.js%20line%205005%20%3E%20eval:7
<anonymous> bundle.js%20line%205005%20%3E%20eval:1
<anonymous> bundle.js:5005
__webpack_require__() bundle.js:556
hotCreateRequire/fn() bundle.js:87
<anonymous> bundle.js%20line%204993%20%3E%20eval:29
<anonymous> bundle.js%20line%204993%20%3E%20eval:1
<anonymous> bundle.js:4993
__webpack_require__() bundle.js:556
hotCreateRequire/fn() bundle.js:87
<anonymous> bundle.js%20line%204081%20%3E%20eval:21
<anonymous> bundle.js:4081
__webpack_require__() bundle.js:556
hotCreateRequire/fn() bundle.js:87
<anonymous> bundle.js%20line%20691%20%3E%20eval:19
<anonymous> bundle.js:691
__webpack_require__() bundle.js:556
hotCreateRequire/fn() bundle.js:87
<anonymous> bundle.js:588
__webpack_require__() bundle.js:556
<anonymous> bundle.js:579
<anonymous> bundle.js:1
1 bundle.js%20line%205005%20%3E%20eval:8312:1

**Comment by nehakhandelwal** [comment]
this is giving the error of 
webworkify-webpack: Could not locate module containing worker function! Make sure you aren't using eval sourcemaps and that you pass named functions to webworkify-webpack!

**Comment by amercier** [comment]
I was able to make it work with webpack using the bundled version (as opposed as the cjs version):

```
import videojs from 'video.js';
window.videojs = videojs; // this is required by the bundled version
require('videojs-contrib-hls/dist/videojs-contrib-hls');
```

With this, no need for any custom config. This works in ES6 too, using `import hls from 'videojs-contrib-hls/dist/videojs-contrib-hls'` intead of `require(...)`.

Versions:
- video.js: 5.11.9
- videojs-contrib-hls: 3.6.6
- webpack: 1.13.3

**Comment by Ambroos** [comment]
If anyone else encounters this issue again (or a similar one) and wants to use videojs-contrib-hls with Webpack:

 `webworkify-webpack` shims window to {}, which causes `global/window` to return the wrong context in a web worker, which in turn causes things to fail. I've created a fork of `webworkify-webpack` that's compatible with the original `webworkify` and works with videojs-contrib-hls!

Just add `webworkify-webpack-dropin@^1.1.9` to your dependencies and add the following alias to your Webpack config:

```javascript
resolve: {
  alias: {
    webworkify: 'webworkify-webpack-dropin',
  },
}
```

It's been tested on videojs-contrib-hls v3.7.0-beta3 and video.js 5.12.6.

**Comment by gkatsev** [comment]
Anyone know of a babel webworkify plugin? If we can inline the webworker at that point, we can easily support both webpack and browserify without extra work on the user's part

**Comment by gkatsev** [comment]
@Ambroos also, thanks for making the dropin replacement package, seems like it'll make things a lot simpler in the meantime.

**Comment by vpowers** [comment]
@Ambroos I've tried your solution and I'm getting an error thrown here in webworkify-webpack-dropin:

```
if (typeof key === 'undefined') {
        throw new Error('webworkify-webpack: Could not locate module containing worker function! Make sure you aren\'t using eval sourcemaps and that you pass named functions to webworkify-webpack!');
}
```
I've made sure that I'm not using eval sourcemaps.  
I've upgraded my versions of videojs-contrib-hls and video.js to v3.7.0-beta3 and 5.12.6, respectively.  The only difference is my project is a react app using nwb, if that makes any difference.  Any ideas?

**Comment by Ambroos** [comment]
@vpowers Could you show me your webpack config? I can't really think of any reason for this error to appear.

**Comment by vpowers** [comment]
@Ambroos Here is my nwb.config.js.  Webpack configuration is passed in:
`module.exports = {
...
  webpack: {
    devtool: 'cheap-source-map',
    html: {
      favicon: 'vendors/favicon.ico'
    },
    extra: {
      resolve: {
        alias: {
          webworkify: 'webworkify-webpack-dropin'
        }
      }
    }
  }
}
`

Here is the line in virtual-source-buffer.js (es5) from videojs-contrib-media-sources that executes webworkify-webpack-dropin:

`// append muxed segments to their respective native buffers as
 // soon as they are available
this.transmuxer_ = (0, _webworkify2['default'])(_transmuxerWorker2['default']);`

**Comment by vpowers** [comment]
I've attempted to isolate the issue by creating a small app using videojs-contrib-hls and bundling with webpack and @Ambroos 's solution does work.  I had been testing this within a larger react app that uses nwb so somethings going on there.  I'll provide an update if I can figure out the issue.

**Comment by dbryand** [comment]
@Ambroos's solution worked for me, but I had to turn off eval for source maps in my large react app, @vpowers.

**Comment by vpowers** [comment]
@dbryand My react app is using nwb for bundling instead of plain webpack.  (nwb uses webpack under the hood).  This seems to be the difference.  Are you just using webpack?

**Comment by dbryand** [comment]
Yes, a home-rolled webpack config.

**Comment by vpowers** [comment]
I've created two small applications. One using webpack and video.js with hls plugin and a react app using nwb and video.js with hls plugin.  Both of them are swapping out the webworkify module with @Ambroos's webworkify-webpack-dropin module.  The webpack app is working with this solution but the react app using nwb is not.  If anyone has any experience with nwb or is feeling helpful, take a look at these projects.  Any help would be appreciated!

https://github.com/vpowers/simple-react-app
https://github.com/vpowers/simple-webpack-app

**Comment by vpowers** [comment]
In case someone comes along and needs to support videojs-contrib-hls in a react app using [nwb](https://github.com/insin/nwb), I have forked @Ambroos `webworkify-webpack-dropin` to support how nwb bundles code:

https://github.com/vpowers/webworkify-nwb

Follow the instructions in this [comment](https://github.com/videojs/videojs-contrib-hls/issues/600#issuecomment-258400412) but use the module above instead.

**Comment by gkatsev** [comment]
I wonder if we can turn webworkify into a babel plugin. Then we could just run it at build time and users of contrib-hls won't need to bother with configuring their bundlers.

**Comment by oleynikd** [comment]
Hi, guys, could please someone explain how to use this inside SPA? I tried @vpowers solution, but I still get `webworkify-webpack: Could not locate module containing worker function! Make sure...` I'm using [vuejs](https://github.com/vuejs/vue)

Thank you in advance.

**Comment by chrispappas** [comment]
Running into this issue here as well, same error as @oleynikd is reporting - any updates since last month on this?

**Comment by lionxcr** [comment]
So Basically here is the setup you need to get this up and running.

I am using node 7.4 and a boiler from create-react-app.

This is my webpack config:

```
resolve: {
        // This allows you to set a fallback for where Webpack should look for modules.
        // We read `NODE_PATH` environment variable in `paths.js` and pass paths here.
        // We use `fallback` instead of `root` because we want `node_modules` to "win"
        // if there any conflicts. This matches Node resolution mechanism.
        // https://github.com/facebookincubator/create-react-app/issues/253
        fallback: paths.nodePaths,
        // These are the reasonable defaults supported by the Node ecosystem.
        // We also include JSX as a common component filename extension to support
        // some tools, although we do not recommend using it, see:
        // https://github.com/facebookincubator/create-react-app/issues/290
        extensions: ['.js', '.json', '.jsx', ''],
        alias: {
            // Support React Native Web
            // https://www.smashingmagazine.com/2016/08/a-glimpse-into-the-future-with-react-native-for-web/
            'react-native': 'react-native-web',
            'webworkify$': 'webworkify-webpack-dropin',
            'videojs-contrib-hls':paths.appNodeModules+'/videojs-contrib-hls/dist/videojs-contrib-hls.min.js'
        }
    }
```

I am also using this plugin to set videojs globally on my scope:

```
new webpack.ProvidePlugin({
            "videojs": "video.js",
            "window.videojs": "video.js"
        }),
```

Additionally on my video component JS I use the following:

```
import videojs from 'video.js';
import 'videojs-contrib-hls';

let video;
```

Later on my component did mount:


```
componentDidMount(){
        const options = {
            height: "100%",
            width: "100%",
            hls: {
                withCredentials: true
            }
        };
        video = videojs("video-player", options);
    }

```

finally you need to set your src m3u8 file and type using Javascript if not this will not work:

```
video.src({
                src:'https://master.m3u8',
                type: 'application/x-mpegURL'
            });
video.play();

render(){
    return(
         <video id="video-player" controls preload className="video-js" poster="videoposter.jpg"/>
     );
}

```

Finally got my video to play hope this helps!

NPM Modules used:

```
"video.js": "^5.12.6",
 "videojs-contrib-hls": "^3.7.0-beta3",
 "webworkify-webpack-dropin": "^1.1.9"
```

**Comment by icetronics** [comment]
Thanks for this @lionxcr, I was able to set it up following your guide.
Is there no other way to integrate it more seamlessly without additional Webpack config?
Thanks!

**Comment by lionxcr** [comment]
@icetronics You are so welcome! As far as I know contrib-hls is not supported in webpack just yet so by far this is the best hack to get m3u8 videos to work properly in react.

**Comment by gkatsev** [comment]
The reason why is because we use browserify to build out the pre-build standalone file. To do that we use the webworkify module that you're replacing with the dropin one for webpack. What would need to happen is to switch to either a babel plugin or something external to the bundler to do the webworker inlining.

**Comment by icetronics** [comment]
Thanks @gkatsev! Would you be able to provide an example or some further pointers so I could try setting it up?

**Comment by cansin** [comment]
In case there are people having the same problems as I did:

I have been trying to make this work with webpack 2.2.1, video.js 5.18.4 and videojs-contrib-hls 5.3.3. I was able to get `webworkify-webpack-dropin` to work with a small change: https://github.com/Ambroos/webworkify-webpack-dropin/pull/1/files . I am hoping @Ambroos will merge the PR and release the version to NPM soon. 

Once that is done all you need to do is:

```
resolve: {
  alias: {
    webworkify: 'webworkify-webpack-dropin',
  },
}
```

at your configuration, and

```
import 'videojs-contrib-hls';
```

at the source code.

**Comment by everett1989** [comment]
@cansin and anyone else who's using webpack's `UglifyJsPlugin`. I found that if I set `compress` to `false` and `mangle` to `true`, like so:

```
plugins: [
    ...
    new webpack.optimize.UglifyJsPlugin({
      compress: false, // leave false for now. This breaks the videojs-contrib-hls package
      mangle: true,
      sourceMap: true
    }),
    ...
],
```
This will uglify it without breaking the `videojs-contrib-hls` package. For me, it was like less than 40KB difference between `compress` being `true` vs `false`. 

I'm also importing the package like so: `import 'videojs-contrib-hls/dist/videojs-contrib-hls.min.js'`

This is a nice workaround for me until `videojs-contrib-hls` is compliant with the compression tool

**Comment by cansin** [comment]
I actually couldn't figure out how to make `webworkify-webpack-dropin` work with my case. Instead I ended up separating videojs-contrib-hls through:

```
require.ensure(['videojs-contrib-hls/dist/videojs-contrib-hls'], require => {
        require('videojs-contrib-hls/dist/videojs-contrib-hls');
}, 'package-videojs-contrib-hls');
```

and then disabled UglifyJS for this separate bundle via:

```
new webpack.optimize.UglifyJsPlugin({
    sourceMap: false,
    compress: {
        warnings: false,
    },
    exclude: /package-videojs-contrib-hls/,
})
```

I know this is less than ideal, yet it works for now.

**Comment by everett1989** [comment]
@cansin I'll try it out, thanks for the example

**Comment by cjdeleon62** [comment]
hi all, does anyone know if this webpack fix works when deploying to heroku? I am currently trying to deploy to heroku and I am getting an 'Uncaught SyntaxError: Unexpected token {'

**Comment by MCDELTAT** [comment]
So does anyone have an actual working HLS + webpack deployment? I'm new to React/Webpack and don't think I can figure out how to get this hack working. Anyone care to create a gist that has all the changes they made to make this work? Appreciate it.

**Comment by thecotne** [comment]
it's not UglifyJs's bug it's bug in this code https://github.com/Ambroos/webworkify-webpack-dropin/blob/master-v1/index.js#L92

it appears when code is minimized and unused function names are removed

```js
              moduleWrapperStrings[moduleId] = wrapperFuncString.substring(0, wrapperFuncString.length - 1) + '\n' + 
fnString.match(/function\s?(.+?)\s?\(.*/)[1] + '();\n}';

```
this regex is matching function name but if there is none it will match some weird shit 

and syntax error will appear after compilation in webworker code generated at runtime  in browser

**Comment by MCDELTAT** [comment]
@thecotne Yeah I had suspected that because for debugging purposes I also moved to Webpack 2, and that didn't fix the issue (has more up to date Uglifyjs) I temporarily solved it by running uglify as part of my "npm build" script outside of webpack. Looks like he just needs to pull from where he forked from because they claim to have a fix that works with Uglify. I'll look at that today.

**Comment by serv** [comment]
When I try to compress the js bundle using webpack and uglify, in the browser console, I get this error

`Uncaught ReferenceError: e is not defined`

**Comment by MCDELTAT** [comment]
@serv Yep. As was discussed in my posts above, there is a very strict regex function in the current version of webworkify-webpack-dropin. Uglify gains one of its shrinkage abilities by deleting functions that it determines are unused, but webworkify is still looking for them. For now, just turn off uglify in webpack, or you can make a npm script that builds your webpack and then runs uglify, which I did with success. I will be trying to get @Ambroos to update his fork here soon, given that I suggested that people use it in the webpack docs.

**Comment by ScottLNorvell** [comment]
Ok... So webworkify-webpack worked for a few weeks to months, then webworkify-webpack-dropin worked for a few more. Now I am on the latest of everything (contrib-hls: 5.4.1, video.js: 5.19.2 as of this writing) and everything is working! Here is how I did it:

Alias videojs-contrib-hls per @lionxcr's suggestion:

```javascript
// webpack.config.js
var path = require('path');
// ...
  resolve: {
    alias: {
      'videojs-contrib-hls': path.join(
        __dirname,
        'node_modules',
        'videojs-contrib-hls',
        'dist',
        // You can use the unminified version and let the minifier minify!
        'videojs-contrib-hls.js'
      ),
    }
  }
// ...
```

Make `videojs` global (also per @lionxcr's suggesiont):

```javascript
// webpack.config.js
// ...
  plugins: [
    new webpack.ProvidePlugin({
        videojs: "video.js",
        "window.videojs": "video.js"
    })
  ]
// ...
```

After all of that, I was getting the same error as @serv. After some digging, I found the [DefinePlugin](https://webpack.js.org/plugins/define-plugin/). It turns out uglify was doing strange things with `global`, so I aliased `typeof global` to `"undefined"` so that the minifier would remove it! 

```javascript
// webpack.config.js
// ...
  plugins: [
    new webpack.DefinePlugin({
      'typeof global': JSON.stringify('undefined')
    })
  ]
// ...
```

Also, if at some point had you aliased `webworkify`, un-alias it!

```javascript
// webpack.config.js
// ...
  resolve: {
    alias: {
      // webworkify$: 'webworkify-webpack',
      // webworkify$: 'webworkify-webpack-dropin',
    }
  }
// ...
```

Phew! That works for now... Hope this helps!

**Comment by correju** [comment]
So far I can create development bundle without a problem and plugin works perfectly, but when I build production I get this error in run-time:

Uncaught ReferenceError: n is not defined blob:http://localhost/659f9283-bbc2-4836-ae79-ebb3cd70a83a:1

I solve this issue passing: 
`devtool: (isProd ? '#eval' : '#source-map')` 
my original set up is 
`devtool: (isProd ? false : '#source-map')`

The problem is that my bundle pass from 1.3MB to 4.3MB when I set eval to devtool.
has anyone solve this issue completely?

**Comment by dbryand** [comment]
The solution by @ScottLNorvell worked for me, but this part led to some very tricky to debug issues with another library (GSAP) in my app. 

```
       new webpack.DefinePlugin({
         'typeof global': JSON.stringify('undefined')
       }),
```

In retrospect it seems obvious that other libs may rely on the `typeof global` and this could blow things up. Looking into alternatives.

**Comment by MCDELTAT** [comment]
?? Decided to visit this again today because I got the classic error 'fs is not defined' in Safari (but not Chrome). But I still can't get webpack to run uglify from the config file. It passes, but then gives like above, 't is not defined in blob' when the page is opened. So I'm going back to my method of just running webpack, and then using uglify after the fact. Somehow that doesn't cause any problems.

**Comment by dbryand** [comment]
Now have confirmed that I'm getting the dreaded 't is not defined', too, when I package and ship to s3. [Example](http://bit.ly/2qALQed)

**Comment by emilioastarita** [comment]
@dbryand same here `"t is not defined"` . Did you find any solution?

**Comment by dbryand** [comment]
Yes, I dropped videojs and just went with html5 + hls.js. Sorry I can't be more helpful :)

**Comment by correju** [comment]
Can you send your webpack configuration?

**Comment by GJordan904** [comment]
After trying everything with the latest stable releases to no avail, I was able to get everything up and running with a modified version of @ScottLNorvell 's solution. 

I used the same versions (contrib-hls: 5.4.1, video.js: 5.19.2) and the following lines are all I have in regards to video.js in my webpack conf:
```javascript
resolve: {
    alias: {
        'videojs-contrib-hls': path.resolve(__dirname, 'node_modules/videojs-contrib-hls/dist/videojs-contrib-hls.js'),
    }
},
plugins: [
    new webpack.ProvidePlugin({
        'window.videojs': 'video.js/dist/video.js',
    }),
]
```
I then simply `require('videojs-contrib-hls');` and what do ya know, no more Code 4 errors and I can read, and most likely play, .m3u8 playlist. 

Only one small problem, my cookies are not being sent with the request. I can generate a signed url and access the playlist, unfortunately, the subsequent requests fail of course so I haven't gotten to test playback yet.

**Comment by correju** [comment]
@GJordan904  I had same problem I solved I solved using this

```
videojs(video, {html5: {
  hls: {
    withCredentials: true
  }
}});
```

**Comment by Adynnor** [comment]
Man this is a bit of a headache.. :( 
Any fix for using uglify? I'm getting these errors
e is not defined
and/or
Unexpected token {

**Comment by VadimBrodsky** [comment]
This Uglify issue was already [fixed](https://github.com/mishoo/UglifyJS2/issues/2101) in Uglify 3.0.17, but Webpack's `uglifyjs-webpack-plugin` [does not support](https://github.com/webpack-contrib/uglifyjs-webpack-plugin/issues/32) Uglify JS 3 yet.

My solution was to downgrade `uglifyjs-webpack-plugin` and `webpack` for now, to versions before the breaking change.

```javascript
// package.json
{
    "uglify-js": "2.7.5",
    "uglifyjs-webpack-plugin": "0.2.2",
    "webpack": "2.2.1"
}
```

**Comment by jou** [comment]
I've convinced videojs-contrib-hls to work with Webpack and Uglify by using the `dist` file and `imports-loader` to define `global` as `undefined`. This causes the check for `global` (which causes the `t is not defined` thingie) to be short circuited and everything seems to be working so far:

```js
// webpack.config.js

resolve: {
    alias: {
        'videojs-contrib-hls': 'videojs-contrib-hls/dist/videojs-contrib-hls',
    }
},
plugins: [
    new webpack.ProvidePlugin({
        'videojs': 'video.js',
        'window.videojs': 'video.js',
    }),
]

// videojs-with-plugins.js (we have a wrapper module that loads video.js plugins and set up the strings we use)

import videojs from 'video.js'

// If you're not using `webpack.ProvidePlugin` anyways, you can also shim `videojs` here direcly
import 'imports-loader?global=>undefined!videojs-contrib-hls'

export default videojs
```

**Comment by zcoding** [comment]
Browserify is dead. Please consider webpack only.

**Comment by jide** [comment]
Folks, there is a much much simpler solution.

Just require the dist bundle like this : 
```js
import 'videojs-contrib-hls/dist/videojs-contrib-hls.js';
```

You may need to make videojs global too : 
```js
import videojs from 'video.js';
window.videojs = videojs;
```

**Comment by jide** [comment]
The main entry in package.json should definitely point at the dist bundle imho.

**Comment by latelt-j** [comment]
@jide I try your solution but I have this error : `Uncaught TypeError: Cannot read property 'EventTarget' of undefined`

**Comment by moshest** [comment]
This is my solution:

```js
// webpack.config.js
{
  resolve: {
    alias: {
      'video.js$': 'video.js/dist/video.cjs.js',
      'videojs-contrib-hls': 'videojs-contrib-hls/dist/videojs-contrib-hls',
    },
  },

  plugins: [
    // Make videojs global
    new webpack.ProvidePlugin({
      videojs: 'video.js',
      'window.videojs': 'video.js',
    }),
  ],
}
```

**Comment by mymainmanbrown** [comment]
@moshest - your answer worked for me - thank you!

Here is the code I used in my actual application after including @moshest webpack config:

```
import videojs from 'video.js';

window.videojs = videojs;

// @see - https://github.com/videojs/videojs-contrib-hls/issues/600#issuecomment-321281442
require('videojs-flash');
require('videojs-contrib-hls');

// ... rest of my application
```

**Comment by mkhazov** [comment]
I've tried every solution described in this issuse but always get `VIDEOJS: ERROR: (CODE:4 MEDIA_ERR_SRC_NOT_SUPPORTED) No compatible source was found for this media` error.

**Comment by markisme** [comment]
So....there is still no some fix is working??

**Comment by AndrewKirkovski** [comment]
I'm getting a different issue
_videoJs.EventTarget is not a constructor in `playlist-loader`

This applies to dist and es5 way of including hls plugin

Making videojs global has no effect, it seems require('video.js') in playlist-loader is returning {default: ...} object instead of videojs itself

Anybody has this problem?

**Comment by jide** [comment]
Oh my... This is driving me nuts. My own solution does not seem to work anymore.

I now have the same issue as @AndrewKirkovski.

There is so much confusion, a dozen solutions here + instructions [here](http://docs.videojs.com/tutorial-webpack.html), each trying to solve a different issue.

Here is a list of the different issues discussed here : 
- webworkify causes uglifyjs error (also makes nwb production build break)
- videojs global
- package.json "main" field not pointing to the dist bundle
- videojs dep returning "default"

Willing to help making sense of all this, but confused atm : 
- Are there compatibility issues with videojs versions here ?
- Why doesn't main point to a bundled file ?
- Why is the videojs global needed ?

**Comment by brainthinks** [comment]
I started getting the same problem that @correju ran into:

`Uncaught ReferenceError: n is not defined`

When I run my project through webpack using uglify.  After some googling, it seems that there is another project, `react-mapbox-gl` has this problem as well.  The issue they came up with is documented here:

https://github.com/mishoo/UglifyJS2/issues/2011

In order to get rid of the error, I had to add the following option to my uglify plugin:

```javascript
{ 
  compress: {
    comparisons: false,
  },
}
```

That option is documented here:

https://github.com/mishoo/UglifyJS2#compress-options

So, all in all, I had to implement the following:

- the `compress` option from this comment (above)
- all options from this comment: https://github.com/videojs/videojs-contrib-hls/issues/600#issuecomment-321281442
- everything from my previous comment: https://github.com/videojs/videojs-contrib-hls/issues/600#issuecomment-321399363

Hope this helps!

**Comment by brainthinks** [comment]
@AndrewKirkovski @jide - are you trying to create a package that utilizes video js, then import your package in another project?

I ask because I got the error you describe in that scenario.  If that is the problem you're running into, here is how I was able to fix it:

```javascript
// webpack.config.js
{
  output: {
      library: {
        root: 'myCustomPlayer',
        amd: 'myCustomPlayer',
        commonjs: 'myCustomPlayer',
      },
      libraryTarget: 'umd',
  },
}
```

See https://webpack.js.org/guides/author-libraries/

**Comment by kenju** [comment]
Hi all,

Would anyone publish the working sample repository? 🙏 

There are so many workarounds here, and I appreciate that. However **the specific versions of libraries (webpack, videojs, videojs-contrib-hls, etc) are also important**, still each workarounds are missing them. That is why a bunch of workarounds have sprung up but no one still finds the best solution.

**Comment by jide** [comment]
For me, the only working combination is @lionxcr answer, with exactly the versions used. Which is sad, beacause I could not make this work with newer releases.

**Comment by jide** [comment]
So, I decided to give https://github.com/benjipott/videojs-hlsjs a try, and it works well. So I thought you should know.

The fun thing is it needs some webpack tricks too, but it was much easier: https://github.com/SRGSSR/videojs-hlsjs/issues/4

And from my early testing, it works better (seeking is faster for example).

**Comment by drexseoj** [comment]
After 2 days of trying different solutions to get this running with webpack and react I finally found a simple solution which does not require any change to the webpack config file.

1. npm install expose-loader --save
2 import videojs and videojs-contrib-hls as follow
```
import "expose-loader?videojs!../node_modules/video.js/dist/video.js";
import "../node_modules/videojs-contrib-hls/dist/videojs-contrib-hls.js";
```

**Comment by gregorskii** [comment]
@jide what does your client code look like with videojs-hlsjs? I tried the details on your linked issue but I have the same code 4 error.

**Comment by jide** [comment]
@gregorskii I just set tech order as `techOrder: ['hlsjs', 'html5', 'youtube']` and import like this :

**Comment by gregorskii** [comment]
@jide Thanks for this. I did not know about the `techOrder`. I added that and I can more clearly see the issue. `videojs-hls` sets the global `videojs` value with the correct techOrder, but it's not available outside of the `videojs-hls` module scope.

My output in the console ends up being this:

<img width="704" alt="screenshot 2017-10-11 15 46 56" src="https://user-images.githubusercontent.com/1130448/31470998-7c9c133c-ae9b-11e7-8010-f867bcc0491c.png">

Inside the `videojs-hlsjs` plugin I consoled `window.videojs`, `window.Hls`, and `window.videojs.options`.

Inside the plugin everything is correct, the tech order gets set, outside its the original videojs that I imported in my module.

I don't want to harp too much on `videojs-hlsjs` as it's not this plugin, but what do you think?

**Comment by gregorskii** [comment]
Similar to @drexseoj this finally appears to be registering:

```
alias: {
      'video.js': 'video.js/dist/video.js',
      'videojs-contrib-hls': 'videojs-contrib-hls/dist/videojs-contrib-hls.js',
    }
```

```
{
  test: /video.js$/,
  use: {
    loader: 'expose-loader',
    query: 'videojs'
  }
},
{
  test: /videojs-contrib-hls$/,
  use: {
    loader: 'imports-loader',
    query: 'video.js'
  }
},
```

```
import videojs from'video.js';
import 'videojs-contrib-hls';
```

```
console.log(window.videojs);
```

<img width="643" alt="screenshot 2017-10-11 16 43 06" src="https://user-images.githubusercontent.com/1130448/31472386-4a15cbee-aea3-11e7-9926-dec986176208.png">

**Comment by jide** [comment]
@gregorskii So everything's fine now ?

**Comment by gregorskii** [comment]
Ya, appears to be working with this plugin.

**Comment by jide** [comment]
If you have errors using ProvidePlugin, be aware that it does not actually expose the module on window, it mimics it, that's why when having some complex setups, exposed modules won't be on window object.

See this comment: https://stackoverflow.com/a/39891365/302731

That's why using expose-loader works in these cases, as @drexseoj said.

**Comment by trinonsense** [comment]
I got a running solution with the uglifyJS issue (`Unexpected token {`):

`package.json`
```
"video.js": "^6.2.8",
"videojs-contrib-hls": "^5.12.0",
"webpack": "1.14.0"
```

`webpack.config.js`
```
// plugins
new webpack.ProvidePlugin({'window.videojs': 'video.js'}), 
new webpack.optimize.UglifyJsPlugin({
  mangle: true,
  comments: false,
  compress: {warnings: false}
})
```

`MyVideo.jsx`
```
import videojs from 'video.js'

// react class
componentDidMount() {
    require('videojs-contrib-hls/dist/videojs-contrib-hls.min')
}
```

**Comment by rparjun** [comment]
Combining inputs from [comment](https://github.com/videojs/videojs-contrib-hls/issues/600#issuecomment-273289790) and [videojs guide](https://github.com/videojs/video.js/blob/master/docs/guides/webpack.md) I was able to make it work with react-babel-webpack project:

webpack: `"^3.4.1"`
babel: `"^6.5.2"`

`yarn add video.js@6.2.8 `
`yarn add videojs-contrib-hls@5.11.0`

Webpack config:
```
plugins = [ 
  new webpack.ProvidePlugin({
    videojs: "video.js",
   "window.videojs": "video.js"
  })
],

module:{
  loaders: [
    {
      test: /\.(png|woff|woff2|eot|ttf|svg)$/,
      loader: 'url-loader?limit=100000',
    }
  ]
}

resolve: {
  alias: {
    webworkify: 'webworkify-webpack-dropin',
    'videojs-contrib-hls': 'videojs-contrib-hls/dist/videojs-contrib-hls.min.js'
  }
}
```

React component:
```
import React from 'react';
require('video.js/dist/video-js.css');
import 'videojs-contrib-hls';

class Videojs extends React.Component{
  componentDidMount() {
    this.player = window.videojs(this.videoNode, {});
  }
  render(){
    return(
      <video ref={ node => this.videoNode = node } controls className="video-js vjs-default-skin">
        {/*
          <source src="/media/arjun/ABC4121123424.mp4" type="video/mp4" />
        */}
        <source src="/media/arjun/31_TEST.m3u8" type="application/x-mpegURL" />
      </video>
    )
  }
}
export default Videojs;
```
We can also [lazy load](https://webpack.js.org/guides/lazy-loading/) this component to the app.

Thanks @lionxcr ! :smile:

**Comment by alexandrzavalii** [comment]
Thanks @rparjun :)
I managed to make it work with your webpack config on AngularJS.
The only mystery remaining is that I have to console.log(HLS) before using it. Yes, sounds very strange :)

```
import * as HLS from 'videojs-contrib-hls';
require('!style-loader!css-loader!video.js/dist/video-js.css');
export function VideoController(...) {
  let player,
       $video = $element.find('video');

function onInit() {
        console.log(HLS); //if I remove this HLS is not loaded.
        const options = {};
        player = videojs($video[0], options);

        player.src({
            src:'https://s3.amazonaws.com/_bc_dml/example-content/bipbop-advanced/bipbop_16x9_variant.m3u8',
            type: 'application/x-mpegURL'
        });
        player.play();

}
```

Any ideas to solve this ?

**Comment by rparjun** [comment]
@alexandrzavalii Please check the edits in my comment, the config may not work in production environment.

**Comment by AndrewKirkovski** [comment]
I've done a lot of digging and it seems there is no reliable solution while `webworkify-webpack-dropin` is used. Problem with that module is that is searches for webpack modules by `toString()` of default export.

Default export can be as simple as `function(t){ new d(t)}` I've encountered situation where 2 modules has exactly same minified string for default export function. So *sometimes* it work, sometimes it doesn't and that depends on how minification behaves and if you use other minified dependencies.

`webworkify-webpack` is using searching by moduleId since v2 and that works reliably. Unfortunately it is no longer a drop in replacement as needs usage modifications i.e. `require.resolve instead of regular require\import`

I've ended up using own hacked `webworkify-webpack-dropin` version that is better aware of what's it is looking for and has function name regexp patched. If I manage to make it less hacky I'll post it somewhere.

**Comment by MCDELTAT** [comment]
For what it's worth, I started working on a new project for a client and wanted to investigate if I needed to change anything up, and I decided to test out the regular 'videojs-contrib-hls' and was getting all of those same fun problems. Instead I followed @rparjun and his comment above to use https://github.com/Peer5/videojs-contrib-hls.js and everything worked right out of the box. I haven't compiled a production webpack build yet, but I'm feeling good about that.

**Comment by rparjun** [comment]
@MCDELTAT I am using the same in some production environments from December and no issues until now. :smiley:

**Comment by mjneil** [comment]
I've got some (hopefully) good news for you all. I've created [webwackify](https://github.com/videojs/webwackify), which is basically the webworkify and webworkify-webpack-dropin projects combined into a single module. I've got an experimental build of videojs-contrib-hls using this project and it can be bundled with browserify and webpack with no extra config options as seen in [this example project](https://github.com/mjneil/sample_videojs_contrib_hls_webwackify).

However, this example is extremely basic so I was hoping some of you could try out this build without the tricks mentioned in this thread and see if it resolves the issues you are having. If successful we can bring this in upstream.

The build has been published under the `qa` tag as version `5.12.3-rc-1`, and you can install with
```
npm install videojs-contrib-hls@qa
```
or you can specify the version directly in `package.json`
```
"videojs-contrib-hls": "5.12.3-rc-1"
```

**Comment by mjneil** [comment]
Has anyone been able to give the build in my previous comment a try? Would really appreciate any feedback on whether it works in any of your custom builds or what breaks.
Pinging a few people that have been active recently @rparjun @MCDELTAT @AndrewKirkovski @brainthinks @jide @gregorskii

**Comment by rparjun** [comment]
@mjneil Thanks :smiley: 
I will take a look into it this weekend and let you know.

**Comment by rparjun** [comment]
Hi @mjneil , `videojs-contrib-hls@qa` is working fine. I was able to verify it will multiple streams, all looks good. :tada:

**Comment by alexandrzavalii** [comment]
@mjneil 
Big thanks! 
Tested thoroughly , all works 
Finally we have a clean solution!

**Comment by kapersoft** [comment]
@mjneil I have tested it with a Laravel Mix / VueJS setup and it works like a charm! Thanks :smiley:

**Comment by mjneil** [comment]
Thank you @rparjun @alexandrzavalii and @kapersoft for testing it out and confirming it works. v5.13.0 of videojs-contrib-hls has been released that include these fixes! Anyone looking to support webpack should upgrade to 5.13.0 and should be able to include the project without extra webpack configuration.

I am closing this issue as resolved. Please reopen the issue or create a new one if you experience any issues after updating.

**Comment by rparjun** [comment]
@mjneil  Thanks :smile:

**Comment by ffischetti** [comment]
@mjneil with v5.13.0 I'm still getting the [unexpected identifier issue](https://github.com/videojs/videojs-contrib-hls/issues/600#issuecomment-340053787)  Anyone else have that problem?

**Comment by kapersoft** [comment]
@ffischetti I have tested v5.13.0 with my Laravel Mix / VueJS setup and didn't experience any problems.

**Comment by thecotne** [comment]
@ffischetti do you import `videojs-contrib-hls/dist/videojs-contrib-hls.min` or `videojs-contrib-hls` ?

i have webpack + react + videojs + videojs-contrib-hls and with latest version everything works fine

do you still have any "hacks" left maybe?

**Comment by ffischetti** [comment]
Thanks @thecotne  

I'm doing: import 'videojs-contrib-hls,' as when I try to import videojs-contrib-hls/dist/videojs-contrib-hls.min, I get the following error: videojs-contrib-hls.min.js:2 Uncaught TypeError: Cannot read property 'EventTarget' of undefined

https://github.com/Peer5/videojs-contrib-hls.js works fine.  But video-js-contrib-hls throws the unexpected token '{' error.

FYI, the app is bootstrapped with create-react-app.

**Comment by thecotne** [comment]
@ffischetti `videojs-contrib-hls.js` is completely different package and what i am saying is that this package `videojs-contrib-hls` works for me and maybe you are doing something wrong

**Comment by thecotne** [comment]
if you import like this 

```js
import 'videojs-contrib-hls'
import videojs from 'video.js'
```

it should work (works for me)

**Comment by ffischetti** [comment]
@thecotne I'm doing exactly that and it throws the error.  I'm aware that videojs-contrib-hls.js is a different package.  I trying to point out that I seem to be experiencing the same state the others did **prior to the fix**.  

I appreciate the help and it's totally possible I'm doing something wrong.  However, I suspect that some other dependency or action might be necessary to get  v5.13.0 working with create-react-app, and that just hasn't been made explicit.  

Out of curiosity, does anyone have this working with create-react-app?  If so, please share any details.  They'd be much appreciated.  Thanks in advance!

**Comment by rparjun** [comment]
@ffischetti I can take a look into this if you can share a repo where i can reproduce this issue.

**Comment by ffischetti** [comment]
Hi @rparjun,

Thanks for the kind offer.  I've created a very basic app that will allow you to reproduce the issue: 
[sample app](https://github.com/ffischetti/hls-test-webpack-debugging).

Please let me knows if this helps or not.  Thanks!

**Comment by mjneil** [comment]
@ffischetti it looks like its the way react activates the css-loader is causing issues, its using the [inline with json object options](https://webpack.js.org/concepts/loaders/#inline) method. Looks like the keys of that json options object have `"` that are not being escaped by webwackify.

**Comment by mjneil** [comment]
@ffischetti this should be fixed in v5.14.0

**Comment by ffischetti** [comment]
Confirmed it's working.  You rock.  Thanks @mjneil !

**Comment by MosheAtOwal** [comment]
Worked for me too, with Webpack (2!) and VueJS.
